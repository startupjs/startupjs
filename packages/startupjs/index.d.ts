export * from '@startupjs/react-sharedb'
export * from '@startupjs/hooks'
// HINT: `isomorphic` means that the code can be executed both
//        on the server and on the client
export * from '@startupjs/isomorphic-helpers'
// dummy babel macro functions for @startupjs/babel-plugin-rn-stylename-inline
export function css (): void
export function styl (): void
