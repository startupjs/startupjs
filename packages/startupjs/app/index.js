module.exports = require('@startupjs/app')
