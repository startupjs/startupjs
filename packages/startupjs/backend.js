export { default } from '@startupjs/backend'
export * from '@startupjs/backend'
