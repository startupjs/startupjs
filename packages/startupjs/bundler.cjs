throw new Error(`
[startupjs/bundler] ERROR! StartupJS bundler api has changed.

Now instead of using general \`require('startupjs/bundler.cjs')\` you have to require
a specific file you need.

Here is what you need to change:

1. \`require('startupjs/bundler.cjs').webpackWebConfig\`
    v v v
   \`require('startupjs/bundler/webpack.web.config.cjs')\`

2. \`require('startupjs/bundler.cjs').webpackServerConfig\`
    v v v
   \`require('startupjs/bundler/webpack.server.config.cjs')\`

3. \`require('startupjs/bundler.cjs').babelConfig\`
    v v v
   \`require('startupjs/bundler/babel.config.cjs')\`

4. \`require('startupjs/bundler.cjs').metroConfig\`
    v v v
   \`require('startupjs/bundler/metro.config.cjs')\`

5. \`require('startupjs/bundler.cjs').rnConfig\`
    v v v
   \`require('startupjs/bundler/react-native.config.cjs')\`

6. \`require('startupjs/bundler.cjs').viteConfig\`
    v v v
   \`require('startupjs/bundler/vite.config.cjs')\`

Additionally, if you did override .js/.jsx compilation in your webpack.web.config.cjs, you
might have used \`getJsxRule\` function there. In which case you should now get it from helpers:

7. \`const { getJsxRule } = require('startupjs/bundler.cjs')\`
    v v v
   \`const { getJsxRule } = require('startupjs/bundler/helpers.cjs')\`
`)
