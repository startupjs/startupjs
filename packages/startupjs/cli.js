#!/usr/bin/env node
'use strict'

import cli from '@startupjs/cli'
import { fileURLToPath } from 'url'
import path, { dirname } from 'path'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

const templatesPath = path.join(__dirname, 'templates')

cli.run({ templatesPath })
