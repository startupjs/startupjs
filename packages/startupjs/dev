#!/bin/sh
# shellcheck disable=SC2059

# Exit on first failed command
set -e

# Run startupjs project using only Docker
#
# Version: 0.1.0
#
# Suports Mac, Linux, Windows WSL2
#
# Usage:
#   ./dev - universal command to run the main terminal or spawn additional terminals.
#           It tries to do the following:
#           1. Run the dev docker image of the same version of startupjs as in package.json.
#           2. If dev image is already running -- exec into it.

# Use current folder name as the name for container and the image by default
# ref: https://stackoverflow.com/a/43919044
a="/$0"; a=${a%/*}; a=${a#/}; a=${a:-.}; SCRIPT_DIR=$(cd "$a"; pwd)
FOLDER_NAME=$( basename "$SCRIPT_DIR" )

IMAGE_NAME="startupjs/dev"
CONTAINER_NAME="$FOLDER_NAME"

# Get startupjs version from package.json
STARTUPJS_VERSION=$( [ -f ./package.json ] && (grep -m 1 '"startupjs"' < ./package.json | sed -n 's/^[^"]*"[^"]*"[^"]*"//p' | sed -n 's/"[^"]*$//p') || echo '' )
STARTUPJS_VERSION=${STARTUPJS_VERSION:-latest}
# Remove leading ^ or ~ and remove the last patch number.
# This way ^0.40.2 will be converted to 0.40
# TODO: if yarn.lock exists, get the exact patch version from there
STARTUPJS_VERSION=$( echo "$STARTUPJS_VERSION" | sed -n 's/^[\^\~]*//p' | sed -n 's/\.*[0-9]*$//p' )

# Helper constants
ROOT="${PWD}"
# folder to store the MongoDB data
DB_FOLDER="${ROOT}/.mongo"
OS_NAME=$( uname -s | tr A-Z a-z )
# on Mac OS we use a faster driver for volume mounts
VOLUME_MODE=$( [ "$OS_NAME" = "darwin" ] && echo ":delegated" || echo "" )

# Colors
COLOR_NONE='\033[0m'
COLOR_LIGHT_GREEN='\033[1;32m'
COLOR_LIGHT_RED='\033[1;31m'

main () {
  if docker ps | grep "\s${CONTAINER_NAME}$"; then
    printf "${COLOR_LIGHT_GREEN}EXEC into the running container '${CONTAINER_NAME}'...${COLOR_NONE}\n"
    _exec
  else
    printf "${COLOR_LIGHT_GREEN}RUN a new dev container '${CONTAINER_NAME}'...${COLOR_NONE}\n"
    _cleanup
    _run
  fi
}

_exec () {
  printf "${COLOR_LIGHT_GREEN}You are now inside the dev environment. Run your commands here:${COLOR_NONE}\n"
  docker exec -it \
    -e "CONTAINER_NAME=${CONTAINER_NAME}" \
    "${CONTAINER_NAME}" \
    bash
}

_run () {
  PORT="${PORT:-$( _find_open_PORT )}"
  DEV_PORT="${DEV_PORT:-$( _find_open_DEV_PORT )}"
  if [ "$PORT" != "3000" ]; then
    printf "${COLOR_LIGHT_RED}IMPORTANT: Default port 3000 was occupied. Your app got assigned the port ${PORT}${COLOR_NONE}!\n"
  fi
  printf "${COLOR_LIGHT_GREEN}You are now inside the dev environment. Run your commands here:${COLOR_NONE}\n"
  echo "CONTAINER_NAME: $CONTAINER_NAME"
  echo "PORT: $PORT"
  echo "DEV_PORT: $DEV_PORT"
  echo "ROOT: $ROOT"
  echo "VOLUME_MODE: $VOLUME_MODE"
  echo "HOME: $HOME"
  echo "DB_FOLDER: $DB_FOLDER"
  echo "IMAGE_NAME: $IMAGE_NAME"
  echo "STARTUPJS_VERSION: $STARTUPJS_VERSION"
  docker run --rm -it \
    --name "${CONTAINER_NAME}" \
    --pull=always \
    -e "CONTAINER_NAME=${CONTAINER_NAME}" \
    -e "DOCKER_HOST_PORT=${PORT}" \
    --ulimit nofile=65535:65535 \
    -p "${PORT}:3000" \
    -p "${DEV_PORT}:3010" \
    -v "${ROOT}:/ws${VOLUME_MODE}" \
    -v "${HOME}/.ssh:/root/.ssh" \
    -v "${HOME}/.gitconfig:/root/.gitconfig" \
    -v "${DB_FOLDER}:/data/db${VOLUME_MODE}" \
    "${IMAGE_NAME}:${STARTUPJS_VERSION}"
}

# Keep only 5 latest startupjs images you pulled. Remove older ones
# ref: https://stackoverflow.com/a/40892314
_cleanup () {
  # shellcheck disable=SC2046
  docker rmi $(docker images -q ${IMAGE_NAME} | tail -n +6) || true
}

# a bit hardcody way to get available port for the app. Start with 3000 and go up to 3009
_find_open_PORT () {
  for iterator in 0 1 2 3 4 5 6 7 8 9; do
    a_port="300${iterator}"
    if ! lsof -n -i tcp:$a_port; then
      echo $a_port
      break
    fi
  done
}

# the same hardcody way as above to get available port for webpack devserver. Start with 3010 and go up to 3019
_find_open_DEV_PORT () {
  for iterator in 0 1 2 3 4 5 6 7 8 9; do
    a_port="301${iterator}"
    if ! lsof -n -i tcp:$a_port; then
      echo $a_port
      break
    fi
  done
}

main "$@"; exit
