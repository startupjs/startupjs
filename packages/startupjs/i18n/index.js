export * from '@startupjs/i18n'
