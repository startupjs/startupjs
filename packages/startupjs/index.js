export * from '@startupjs/react-sharedb'
export * from '@startupjs/hooks'
// HINT: `isomorphic` means that the code can be executed both
//        on the server and on the client
export * from '@startupjs/isomorphic-helpers'
// dummy babel macro functions for @startupjs/babel-plugin-rn-stylename-inline
export function css () {}
export function styl () {}
export { default as t } from '@startupjs/i18n/client/t.js'
