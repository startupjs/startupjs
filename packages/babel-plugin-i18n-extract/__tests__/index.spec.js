const pluginTester = require('babel-plugin-tester').default
const plugin = require('../index')
const { name: pluginName } = require('../package.json')

// TODO: Add test that tests merge extensions

pluginTester({
  plugin,
  pluginName,
  snapshot: true,
  pluginOptions: {},
  babelOptions: {
    plugins: ['@babel/plugin-syntax-jsx']
  },
  // since the filename is empty when we pass the 'code' to test object
  // then we could use fixtures
  // but we can't use fixtures because of this bug https://github.com/babel-utils/babel-plugin-tester/issues/74
  // so as a workaround just pass the filename in the plugin options
  tests: [
    {
      title: 'Call without arguments',
      error: true,
      code: `
        import { observer, t } from 'startupjs'

        export default observer(function App () {
          return (
            <span>
              {t()}
            </span>
          )
        })
      `
    },
    {
      title: 'Call only with \'key\' argument',
      error: true,
      code: `
        import { observer, t } from 'startupjs'

        export default observer(function App () {
          return (
            <span>
              {t('key')}
            </span>
          )
        })
      `
    },
    {
      title: 'Call with empty \'key\' argument',
      error: true,
      code: `
        import { observer, t } from 'startupjs'

        export default observer(function App () {
          return (
            <span>
              {t('', 'defaultValue')}
            </span>
          )
        })
      `
    },
    {
      title: 'Call with empty \'defaultValue\' argument',
      error: true,
      code: `
        import { observer, t } from 'startupjs'

        export default observer(function App () {
          return (
            <span>
              {t('key', '')}
            </span>
          )
        })
      `
    },
    {
      title: 'String literal is not passed to the \'key\' argument',
      error: true,
      code: `
        import { observer, t } from 'startupjs'
        const key = 'key'

        export default observer(function App () {
          return (
            <span>
              {t(key, 'defaultValue')}
            </span>
          )
        })
      `
    },
    {
      title: 'String literal is not passed to the \'defaultValue\' argument',
      error: true,
      code: `
        import { observer, t } from 'startupjs'
        const defaultValue = 'defaultValue'

        export default observer(function App () {
          return (
            <span>
              {t('key', defaultValue)}
            </span>
          )
        })
      `
    },
    {
      title: 'Call with custom name',
      error: true,
      code: `
        import { observer, t as myT } from 'startupjs'
        const key = 'key'

        export default observer(function App () {
          return (
            <span>
              {myT(key, 'defaultValue')}
            </span>
          )
        })
      `
    },
    {
      title: 'Correct call',
      pluginOptions: {
        filename: 'correct.js'
      },
      code: `
        import { observer, t } from 'startupjs'

        export default observer(function App () {
          return (
            <span>
              {t('key', 'defaultValue')}
            </span>
          )
        })
      `
    },
    {
      title: 'Ignore if not import',
      code: `
        import { observer } from 'startupjs'
        export default observer(function App () {
          const t = useTranslation()
          return (
            <span>
              {t('key', 'defaultValue')}
            </span>
          )
        })
      `
    },
    {
      title: 'Ignore shadowing',
      code: `
        import { observer, t } from 'startupjs'
        export default observer(function App () {
          const t = useTranslation()
          return (
            <span>
              {t('key', 'defaultValue')}
            </span>
          )
        })
      `
    },
    {
      title: 'Ignore if not used import with custom name',
      code: `
        import { observer, t as myT } from 'startupjs'
        export default observer(function App () {
          const t = useTranslation()
          return (
            <span>
              {t('key', 'defaultValue')}
            </span>
          )
        })
      `
    }
  ]
})
