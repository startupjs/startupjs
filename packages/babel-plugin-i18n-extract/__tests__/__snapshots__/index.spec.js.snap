// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@startupjs/babel-plugin-i18n-extract Call only with 'key' argument: Call only with 'key' argument 1`] = `

import { observer, t } from 'startupjs'

export default observer(function App () {
  return (
    <span>
      {t('key')}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

Error: unknown: [@startupjs/babel-plugin-i18n-extract]: Not enough arguments passed to the function t

`;

exports[`@startupjs/babel-plugin-i18n-extract Call with custom name: Call with custom name 1`] = `

import { observer, t as myT } from 'startupjs'
const key = 'key'

export default observer(function App () {
  return (
    <span>
      {myT(key, 'defaultValue')}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

Error: unknown: [@startupjs/babel-plugin-i18n-extract]: Argument 'key' of myT should be a string literal

`;

exports[`@startupjs/babel-plugin-i18n-extract Call with empty 'defaultValue' argument: Call with empty 'defaultValue' argument 1`] = `

import { observer, t } from 'startupjs'

export default observer(function App () {
  return (
    <span>
      {t('key', '')}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

Error: unknown: [@startupjs/babel-plugin-i18n-extract]: Argument 'defaultValue' of t cannot be an empty string

`;

exports[`@startupjs/babel-plugin-i18n-extract Call with empty 'key' argument: Call with empty 'key' argument 1`] = `

import { observer, t } from 'startupjs'

export default observer(function App () {
  return (
    <span>
      {t('', 'defaultValue')}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

Error: unknown: [@startupjs/babel-plugin-i18n-extract]: Argument 'key' of t cannot be an empty string

`;

exports[`@startupjs/babel-plugin-i18n-extract Call without arguments: Call without arguments 1`] = `

import { observer, t } from 'startupjs'

export default observer(function App () {
  return (
    <span>
      {t()}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

Error: unknown: [@startupjs/babel-plugin-i18n-extract]: Not enough arguments passed to the function t

`;

exports[`@startupjs/babel-plugin-i18n-extract Correct call: Correct call 1`] = `

import { observer, t } from 'startupjs'

export default observer(function App () {
  return (
    <span>
      {t('key', 'defaultValue')}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

import { observer, t } from "startupjs";
export default observer(function App() {
  return <span>{t("correct%2Ejs.key", "defaultValue")}</span>;
});


`;

exports[`@startupjs/babel-plugin-i18n-extract Ignore if not import: Ignore if not import 1`] = `

import { observer } from 'startupjs'
export default observer(function App () {
  const t = useTranslation()
  return (
    <span>
      {t('key', 'defaultValue')}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

import { observer } from "startupjs";
export default observer(function App() {
  const t = useTranslation();
  return <span>{t("key", "defaultValue")}</span>;
});


`;

exports[`@startupjs/babel-plugin-i18n-extract Ignore if not used import with custom name: Ignore if not used import with custom name 1`] = `

import { observer, t as myT } from 'startupjs'
export default observer(function App () {
  const t = useTranslation()
  return (
    <span>
      {t('key', 'defaultValue')}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

import { observer, t as myT } from "startupjs";
export default observer(function App() {
  const t = useTranslation();
  return <span>{t("key", "defaultValue")}</span>;
});


`;

exports[`@startupjs/babel-plugin-i18n-extract Ignore shadowing: Ignore shadowing 1`] = `

import { observer, t } from 'startupjs'
export default observer(function App () {
  const t = useTranslation()
  return (
    <span>
      {t('key', 'defaultValue')}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

import { observer, t } from "startupjs";
export default observer(function App() {
  const t = useTranslation();
  return <span>{t("key", "defaultValue")}</span>;
});


`;

exports[`@startupjs/babel-plugin-i18n-extract String literal is not passed to the 'defaultValue' argument: String literal is not passed to the 'defaultValue' argument 1`] = `

import { observer, t } from 'startupjs'
const defaultValue = 'defaultValue'

export default observer(function App () {
  return (
    <span>
      {t('key', defaultValue)}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

Error: unknown: [@startupjs/babel-plugin-i18n-extract]: Argument 'defaultValue' of t should be a string literal

`;

exports[`@startupjs/babel-plugin-i18n-extract String literal is not passed to the 'key' argument: String literal is not passed to the 'key' argument 1`] = `

import { observer, t } from 'startupjs'
const key = 'key'

export default observer(function App () {
  return (
    <span>
      {t(key, 'defaultValue')}
    </span>
  )
})

      ↓ ↓ ↓ ↓ ↓ ↓

Error: unknown: [@startupjs/babel-plugin-i18n-extract]: Argument 'key' of t should be a string literal

`;
